Source/Core/VideoCommon/Resources/TextureAndSamplerResource.cpp

```diff
@@ -0,0 +1,105 @@
+// Copyright 2025 Dolphin Emulator Project
+// SPDX-License-Identifier: GPL-2.0-or-later
+
+#include "VideoCommon/Resources/TextureAndSamplerResource.h"
+
+#include "VideoCommon/Assets/CustomAssetCache.h"
+#include "VideoCommon/Resources/TexturePool.h"
+
+namespace VideoCommon
+{
+TextureAndSamplerResource::TextureAndSamplerResource(Resource::ResourceContext resource_context)
+    : Resource(std::move(resource_context))
+{
+  m_texture_and_sampler_asset = m_resource_context.asset_cache->CreateAsset<TextureAndSamplerAsset>(
+      m_resource_context.primary_asset_id, m_resource_context.asset_library, this);
+}
+
+void TextureAndSamplerResource::MarkAsActive()
+{
+  m_resource_context.asset_cache->MarkAssetActive(m_texture_and_sampler_asset);
+}
+
+void TextureAndSamplerResource::MarkAsPending()
+{
+  m_resource_context.asset_cache->MarkAssetPending(m_texture_and_sampler_asset);
+}
+
+const std::shared_ptr<TextureAndSamplerResource::Data>& TextureAndSamplerResource::GetData() const
+{
+  return m_current_data;
+}
+
+void TextureAndSamplerResource::ResetData()
+{
+  m_load_data = std::make_shared<Data>();
+}
+
+Resource::TaskComplete TextureAndSamplerResource::CollectPrimaryData()
+{
+  m_load_data->m_texture_and_sampler_data = m_texture_and_sampler_asset->GetData();
+  if (!m_load_data->m_texture_and_sampler_data)
+    return Resource::TaskComplete::No;
+
+  auto& texture_data = m_load_data->m_texture_and_sampler_data->texture_data;
+  if (texture_data.m_slices.empty())
+    return Resource::TaskComplete::Error;
+
+  if (texture_data.m_slices[0].m_levels.empty())
+    return Resource::TaskComplete::Error;
+
+  const auto& first_level = texture_data.m_slices[0].m_levels[0];
+
+  auto& config = m_load_data->m_config;
+  config.format = first_level.format;
+  config.flags = 0;
+  config.layers = 1;
+  config.levels = 1;
+  config.type = m_load_data->m_texture_and_sampler_data->type;
+  config.samples = 1;
+
+  config.width = first_level.width;
+  config.height = first_level.height;
+
+  return Resource::TaskComplete::Yes;
+}
+
+Resource::TaskComplete TextureAndSamplerResource::ProcessData()
+{
+  if (auto texture = m_resource_context.texture_pool->AllocateTexture(m_load_data->m_config))
```

Comment:
If you invert this check you can `return Resource::TaskComplete::Error;` here and flatten the rest of the function one level.